textattack/models/wrappers/huggingface_model_wrapper.py

```diff
@@ -24,32 +47,71 @@ def __call__(self, text_input_list):
-            }
-            outputs = self.model(**input_dict)
-
-            if isinstance(outputs[0], str):
-                # HuggingFace sequence-to-sequence models return a list of
-                # string predictions as output. In this case, return the full
-                # list of outputs.
-                return outputs
-            else:
-                # HuggingFace classification models return a tuple as output
-                # where the first item in the tuple corresponds to the list of
-                # scores for each input.
-                return outputs[0]
-
         with torch.no_grad():
             outputs = textattack.shared.utils.batch_model_predict(
-                model_predict, ids, batch_size=self.batch_size
+                self._model_predict, ids, batch_size=self.batch_size
             )
 
         return outputs
+
+    def get_grad(self, text_input):
+        """
+        Get gradient with respect to input tokens
+        Args:
+            text_input (str): input string
+        Returns:
+            Dict of ids, tokens, and gradient as numpy array.
+        """
+        if isinstance(self.model, textattack.models.helpers.T5ForTextToText):
+            raise NotImplementedError(
+                "`get_grads` for T5FotTextToText has not been implemented yet."
+            )
+
+        self.model.train()
+        embedding_layer = self.model.get_input_embeddings()
+        original_state = embedding_layer.weight.requires_grad
+        embedding_layer.weight.requires_grad = True
+
+        emb_grads = []
+
+        def grad_hook(module, grad_in, grad_out):
+            emb_grads.append(grad_out[0])
+
+        emb_hook = embedding_layer.register_backward_hook(grad_hook)
+
+        self.model.zero_grad()
+        model_device = next(self.model.parameters()).device
+        ids = self.tokenize([text_input])
+        if hasattr(self.tokenizer, "convert_ids_to_tokens"):
+            tokens = self.tokenizer.convert_ids_to_tokens(ids[0]["input_ids"])
+        else:
+            tokens = None
+        predictions = self._model_predict(ids)
+
+        model_device = next(self.model.parameters()).device
+        input_dict = {k: [_dict[k] for _dict in ids] for k in ids[0]}
+        input_dict = {
+            k: torch.tensor(v).to(model_device) for k, v in input_dict.items()
+        }
+        try:
+            labels = predictions.argmax(dim=1)
+            loss = self.model(**input_dict, labels=labels)[0]
+        except TypeError:
+            raise TypeError(
+                f"{type(self.model)} class does not take in `labels` to calculate loss. "
+                "One cause for this might be if you instatiated your model using `transformer.AutoModel` "
+                "(instead of `transformers.AutoModelForSequenceClassification`)."
+            )
+
+        loss.backward()
+
+        # grad w.r.t to word embeddings
+        grad = emb_grads[0][0].cpu().numpy()
```

Comment:
sorry, why is this `emb_grads[0][0]` instead of `emb_grads[0]`? I haven't used Hooks in awhile.

Reply:
I believe it's because of input technically a batch of size 1. But I'll double check.

Reply:
Confirmed that it's because of batch size.